zookeeper/dataset.py

```diff
@@ -103,26 +133,77 @@ def load(self, split, decoders, shuffle) -> tf.data.Dataset:
             as_dataset_kwargs={"shuffle_files": shuffle},
         )
 
-    def train_data(self, decoders=None) -> tf.data.Dataset:
-        return self.load(self.train_split, decoders=decoders, shuffle=True)
-
-    @property
-    def num_train_examples(self) -> int:
-        return self.num_examples(self.train_split)
+    def train(self, decoders=None) -> Tuple[tf.data.Dataset, int]:
+        return (
+            self.load(self.train_split, decoders=decoders, shuffle=True),
+            self.num_examples(self.train_split),
+        )
 
-    def validation_data(self, decoders=None) -> tf.data.Dataset:
+    def validation(self, decoders=None) -> Tuple[tf.data.Dataset, int]:
         if self.validation_split is None:
             raise ValueError(
                 f"Dataset {self.__class__.__name__} is not configured with a "
                 "validation split."
             )
-        return self.load(self.validation_split, decoders=decoders, shuffle=False)
+        return (
+            self.load(self.validation_split, decoders=decoders, shuffle=False),
+            self.num_examples(self.validation_split),
+        )
 
-    @property
-    def num_validation_examples(self) -> int:
-        if self.validation_split is None:
-            raise ValueError(
-                f"Dataset {self.__class__.__name__} is not configured with a "
-                "validation split."
+
+class MultiTFDSDataset(Dataset):
+    """
+    A wrapper around multiple TensorFlowDatasets datasets. This allows a model
+    to be trained on data that is combined from multiple datasets.
+    """
+
+    # A non-empty mapping from dataset names as keys to splits as values. The
+    # training data will be the concatenation of the datasets loaded from each
+    # (key, value) pair.
+    train_splits: Dict[str, str]
+
+    # As above, a mapping from dataset names as keys to splits as values. May be
+    # empty, indicating no validation data.
+    validation_splits: Dict[str, str] = {}
+
+    # The directory that the dataset is stored in.
+    data_dir: Optional[str] = None
+
+    def num_examples(self, splits) -> int:
+        """
+        Compute the total number of examples in the splits specified by the
+        dictionary `splits`.
+        """
+
+        return sum(
+            sum(
```

Comment:
I don't think we need this sum, we can just use a double loop, or am I missing something?

Reply:
So I did that because your suggestion didn't work. I just discovered why though, and found a solution.

Apparently this syntax is not allowed
```python
(stmt_in_terms_of_x for x in iterator_in_terms_of_y for y in iterator)
```
but swapping them to this is fine
```python
(stmt_in_terms_of_x for y in iterator for x in iterator_in_terms_of_y)
```
so I'll replace it with
```python
        return sum(
            tfds.builder(name, data_dir=self.data_dir).info.splits[s].num_examples
            for name, split in splits
            for s in base_splits(split)
        )
```

Reply:
(To me the second syntax looks way more confusing but I guess that's Python for you....)

Reply:
Makes sense, sorry didn't actually test my suggestion